src/native/libs/System.Globalization.Native/pal_casing.m

```diff
@@ -61,15 +102,23 @@ int32_t GlobalizationNative_ChangeCaseNative(const uint16_t* localeName, int32_t
         NSString *locName = [NSString stringWithCharacters: localeName length: lNameLength];
         currentLocale = [NSLocale localeWithLocaleIdentifier:locName];
     }
-    NSString *source = [NSString stringWithCharacters: lpSrc length: cwSrcLength];
-    NSString *result = bToUpper ? [source uppercaseStringWithLocale:currentLocale] : [source lowercaseStringWithLocale:currentLocale];
 
     int32_t srcIdx = 0, dstIdx = 0, isError = 0;
     uint16_t dstCodepoint;
-    while (srcIdx < result.length)
+    while (srcIdx < cwSrcLength)
     {
-        dstCodepoint = [result characterAtIndex:srcIdx++];
-        Append(lpDst, dstIdx, cwDstLength, dstCodepoint, isError);
+        int32_t startIndex = srcIdx;
+        NEXTOFFSET(lpSrc, srcIdx, cwSrcLength);
+        int32_t srcLength = srcIdx - startIndex;
+        NSString *src = [NSString stringWithCharacters: lpSrc + startIndex length: srcLength];
+        NSString *dst = bToUpper ? [src uppercaseStringWithLocale:currentLocale] : [src lowercaseStringWithLocale:currentLocale];
+        int32_t index = 0;
+        while (index < srcLength)
+        {
+            dstCodepoint = dst.length > srcLength ? [src characterAtIndex: index] : [dst characterAtIndex: index];
+            Append(lpDst, dstIdx, cwDstLength, dstCodepoint, isError);
+            index++;
+        }
```

Comment:
I think we should note in the description above this function that there are two things we are considering when changing the case of a string here. Something like the below

1) Prohibiting code point expansions. Some characters code points expand when uppercased or lowercased, which may lead to an insufficient destination buffer. Instead, we prohibit these expansions and iterate through the string character by character opting for the original character if it would have been expanded.

2) Properly handling surrogate pairs. Characters can be comprised of more than one code point (i.e. surrogate pairs like `\uD801\uDC37`). All code points for a character are needed to properly change case

And then we can have inline comments to indicate where these considerations are taken, like inner while loop is to iterate over all code points of a surrogate pair character and the `dst.length > srcLength` is to prevent code point expansions